pkg/certmaker/template.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Package certmaker provides template parsing and certificate generation functionality
+// for creating X.509 certificates from JSON templates per RFC3161 standards.
+package certmaker
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"os"
+	"time"
+)
+
+// CertificateTemplate defines the structure for the JSON certificate templates
+type CertificateTemplate struct {
+	Subject struct {
+		Country            []string `json:"country,omitempty"`
+		Organization       []string `json:"organization,omitempty"`
+		OrganizationalUnit []string `json:"organizationalUnit,omitempty"`
+		CommonName         string   `json:"commonName"`
+	} `json:"subject"`
+	Issuer struct {
+		CommonName string `json:"commonName"`
+	} `json:"issuer"`
+	NotBefore        string   `json:"notBefore"`
+	NotAfter         string   `json:"notAfter"`
+	KeyUsage         []string `json:"keyUsage"`
+	BasicConstraints struct {
+		IsCA       bool `json:"isCA"`
+		MaxPathLen int  `json:"maxPathLen"`
+	} `json:"basicConstraints"`
+	Extensions []struct {
+		ID       string `json:"id"`
+		Critical bool   `json:"critical"`
+		Value    string `json:"value"`
+	} `json:"extensions,omitempty"`
+	ExtKeyUsage []string `json:"extKeyUsage,omitempty"`
+}
+
+// ParseTemplate creates an x509 certificate from JSON template
+func ParseTemplate(filename string, parent *x509.Certificate) (*x509.Certificate, error) {
+	content, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, fmt.Errorf("error reading template file: %w", err)
+	}
+
+	var tmpl CertificateTemplate
+	if err := json.Unmarshal(content, &tmpl); err != nil {
+		return nil, fmt.Errorf("error parsing template JSON: %w", err)
+	}
+
+	certType := "root"
+	if parent != nil {
+		if tmpl.BasicConstraints.IsCA {
+			certType = "intermediate"
+		} else {
+			certType = "leaf"
+		}
+	}
+
+	if err := ValidateTemplate(&tmpl, parent, certType); err != nil {
+		return nil, err
+	}
+
+	return CreateCertificateFromTemplate(&tmpl, parent)
+}
+
+// ValidateTemplate performs validation checks on the certificate template.
+func ValidateTemplate(tmpl *CertificateTemplate, parent *x509.Certificate, certType string) error {
+	if tmpl.NotBefore == "" || tmpl.NotAfter == "" {
```

Comment:
I'd slightly prefer an option to set the lifetime and use the current time rather than require exact timestamps.